src/Shared/TryParseMethodCache.cs

```diff
@@ -87,32 +115,59 @@ public static bool HasTryParseMethod(ParameterInfo parameter)
                 return null;
             }
 
-            return MethodCallCache.GetOrAdd(type, Finder);
+            return _methodCallCache.GetOrAdd(type, Finder);
         }
 
-        private static MethodInfo GetEnumTryParseMethod()
+        private static MethodInfo GetEnumTryParseMethod(bool preferNonGenericEnumParseOverload)
         {
             var staticEnumMethods = typeof(Enum).GetMethods(BindingFlags.Public | BindingFlags.Static);
 
+            // With NativeAOT, if there's no static usage of Enum.TryParse<T>, it will be removed
+            // we fallback to the non-generic version if that is the case
+            MethodInfo? genericCandidate = null;
+            MethodInfo? nonGenericCandidate = null;
+
             foreach (var method in staticEnumMethods)
             {
-                if (!method.IsGenericMethod || method.Name != nameof(Enum.TryParse) || method.ReturnType != typeof(bool))
+                if (method.Name != nameof(Enum.TryParse) || method.ReturnType != typeof(bool))
                 {
                     continue;
                 }
 
                 var tryParseParameters = method.GetParameters();
 
-                if (tryParseParameters.Length == 2 &&
+                // Enum.TryParse<T>(string, out object)
+                if (method.IsGenericMethod &&
+                    tryParseParameters.Length == 2 &&
                     tryParseParameters[0].ParameterType == typeof(string) &&
                     tryParseParameters[1].IsOut)
                 {
-                    return method;
+                    genericCandidate = method;
                 }
+
+                // Enum.TryParse(type, string, out object)
+                if (!method.IsGenericMethod &&
+                    tryParseParameters.Length == 3 &&
+                    tryParseParameters[0].ParameterType == typeof(Type) &&
+                    tryParseParameters[1].ParameterType == typeof(string) &&
+                    tryParseParameters[2].IsOut)
+                {
+                    nonGenericCandidate = method;
+                }
+            }
+
+            if (genericCandidate is null && nonGenericCandidate is null)
+            {
+                Debug.Fail("No suitable System.Enum.TryParse method not found.");
+                throw new Exception("No suitable System.Enum.TryParse method not found.");
```

Comment:
Nit: double-negatives.

```suggestion
                Debug.Fail("No suitable System.Enum.TryParse method found.");
                throw new Exception("No suitable System.Enum.TryParse method found.");
```

Reply:
Good catch!